File: src/VisualStudio/IntegrationTest/New.IntegrationTests/AbstractAutomationObjectTests.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Security;
using System.Text;
using System.Threading.Tasks;
using Microsoft.VisualBasic;
using System.Collections.Immutable;
using System.Windows;
using System.Windows.Controls;

using Microsoft.CodeAnalysis.Options;
using Microsoft.VisualStudio.LanguageServices.Implementation.Options;
using Microsoft.VisualStudio.IntegrationTest.Utilities;
using Roslyn.Test.Utilities;
using Xunit;
using Roslyn.VisualStudio.IntegrationTests;
using Microsoft.VisualStudio.Extensibility.Testing;
using Microsoft.CodeAnalysis;
using Microsoft.VisualStudio.Shell;
using Microsoft.VisualStudio.LanguageServices;
using Roslyn.Utilities;

namespace Roslyn.VisualStudio.NewIntegrationTests
{
    public abstract class AbstractAutomationObjectTests : AbstractEditorTest
    {
        protected AbstractAutomationObjectTests() : base(nameof(AbstractAutomationObjectTests))
        {
        }

        [IdeFact]
        public async Task TestOptionsInUIShouldBeInAutomationObject()
        {
            var package = await TestServices.Editor.GetLanguagePackageAsync(LanguageName, HangMitigatingCancellationToken);
            var workspace = await TestServices.Shell.GetComponentModelServiceAsync<VisualStudioWorkspace>(HangMitigatingCancellationToken);
            var automationObject = CreateAutomationObject(workspace);
            foreach (var attribute in package.GetType().GetCustomAttributes(typeof(ProvideLanguageEditorOptionPageAttribute)))
            {
                var pageType = ((ProvideLanguageEditorOptionPageAttribute)attribute).PageType;
                var dialogPage = (AbstractOptionPage)package.GetDialogPage(pageType);
                var optionPageControl = (AbstractOptionPageControl)dialogPage.GetType().GetProperty("Child", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(dialogPage, null);
                var radioButtonGroups = new Dictionary<string, List<RadioButton>>();
                // TODO: Not all options (and even pages) use BindToOption.
                // So the validation here is incomplete.
                foreach (var bindingExpression in optionPageControl.BindingExpressions)
                {
                    var target = (FrameworkElement)bindingExpression.Target;
                    var optionForAssertMessage = target.Name;

                    if (target is CheckBox checkBox)
                    {
                        VerifySingleChangeWhenOptionChangeInUI(dialogPage, automationObject, () => checkBox.IsChecked = !checkBox.IsChecked, optionForAssertMessage);
                    }
                    else if (target is ComboBox comboBox)
                    {
                        VerifySingleChangeWhenOptionChangeInUI(dialogPage, automationObject, () => comboBox.SelectedIndex = comboBox.SelectedIndex == 0 ? 1 : 0, optionForAssertMessage);
                    }
                    else if (target is RadioButton radioButton)
                    {
                        if (radioButtonGroups.TryGetValue(radioButton.GroupName, out var list))
                        {
                            list.Add(radioButton);
                        }
                        else
                        {
                            radioButtonGroups.Add(radioButton.GroupName, new List<RadioButton> { radioButton });
                        }
                    }
                    else
                    {
                        throw ExceptionUtilities.UnexpectedValue(target);
                    }
                }

                foreach (var radioButtonGroup in radioButtonGroups)
                {
                    var groupName = radioButtonGroup.Key;
                    var radioButtons = radioButtonGroup.Value;
                    // There is no point in having a single radio button in a group.
                    Assert.True(radioButtons.Count > 1, $"Expected radio button group '{groupName}' to have more than one radio button. Found {radioButtons.Count}.");
                    var selectedRadioButton = radioButtons.SingleOrDefault(r => r.IsChecked == true);
                    foreach (var radioButton in radioButtons)
                    {
                        // We test selecting every radio button in the group.
                        // We skip the already selected one till we are sure we tested other radio buttons.
                        if (radioButton == selectedRadioButton)
                        {
                            continue;
                        }

                        Assert.False(radioButton.IsChecked);
                        VerifySingleChangeWhenOptionChangeInUI(dialogPage, automationObject, () => radioButton.IsChecked = true, optionForAssertMessage: radioButton.Name);
                    }

                    // TODO: Consider asserting a non-null selectedRadioButton if https://github.com/dotnet/roslyn/issues/62363 is fixed.

                    if (selectedRadioButton is not null)
                    {
                        // Now that we tested other radio buttons in the group, the initially selected radio button is now not selected.
                        Assert.False(selectedRadioButton.IsChecked);
                        VerifySingleChangeWhenOptionChangeInUI(dialogPage, automationObject, () => selectedRadioButton.IsChecked = true, optionForAssertMessage: selectedRadioButton.Name);
                    }
                }
            }

            // Above checks that all options are in AutomationObjects.
            // TODO: check that all automation object members are in options.
        }

        protected abstract AbstractAutomationObject CreateAutomationObject(Workspace workspace);

        private static void VerifySingleChangeWhenOptionChangeInUI(AbstractOptionPage page, AbstractAutomationObject automationObject, Action changeUIControl, string optionForAssertMessage)
        {
            var automationValuesBeforeChange = GetAutomationDictionary(automationObject);

            changeUIControl();

            page.SaveSettingsToStorage();

            var automationValuesAfterChange = GetAutomationDictionary(automationObject);

            Assert.Equal(automationValuesBeforeChange.Count, automationValuesAfterChange.Count);
            AssertExactlyOneChange(automationValuesBeforeChange, automationValuesAfterChange, optionForAssertMessage);
        }

        private static ImmutableDictionary<string, object> GetAutomationDictionary(AbstractAutomationObject automationObject)
        {
            var automationObjectValues = ImmutableDictionary.CreateBuilder<string, object>();
            foreach (var property in automationObject.GetType().GetProperties())
            {
                automationObjectValues.Add(property.Name, property.GetValue(automationObject, null));
            }

            return automationObjectValues.ToImmutable();
        }

        private static void AssertExactlyOneChange(ImmutableDictionary<string, object> dictionary1, ImmutableDictionary<string, object> dictionary2, string optionForAssertMessage)
        {
            var seenChange = false;
            string changedKey = null!;
            foreach (var key in dictionary1.Keys)
            {
                if (!dictionary1[key].Equals(dictionary2[key]))
                {
                    if (seenChange)
                    {
                        if (key != "ShowSnippets" && changedKey != "ShowSnippets" &&
                            key != "EnterKeyBehavior" && changedKey != "EnterKeyBehavior")
                        {
                            Assert.False(true, $"Two values ('{key}' and '{changedKey}') have changed in automation object after changing '{optionForAssertMessage}'!");
                        }
                    }

                    changedKey = key;
                    seenChange = true;
                }
            }

            Assert.True(seenChange, $"No change was found in automation object after changing '{optionForAssertMessage}'.");
        }
    }
}

File: src/VisualStudio/TestUtilities2/Options/AbstractAutomationObjectTests.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.

Imports System.Collections.Immutable
Imports System.Composition
Imports System.Threading
Imports System.Windows
Imports System.Windows.Controls
Imports Microsoft.CodeAnalysis.ColorSchemes
Imports Microsoft.CodeAnalysis.Editor.UnitTests
Imports Microsoft.CodeAnalysis.Editor.UnitTests.Workspaces
Imports Microsoft.CodeAnalysis.Host.Mef
Imports Microsoft.CodeAnalysis.Options
Imports Microsoft.CodeAnalysis.Test.Utilities
Imports Microsoft.VisualStudio.LanguageServices.Implementation.Options
Imports Roslyn.Test.Utilities

Namespace Microsoft.VisualStudio.LanguageServices.UnitTests
    Public MustInherit Class AbstractAutomationObjectTests
        Protected MustOverride ReadOnly Property SkippedOptionsForOptionChangedTest As ImmutableArray(Of String)
        Protected MustOverride ReadOnly Property SkippedOptionsForProperStorageTest As ImmutableArray(Of String)

        <Fact>
        Public Sub TestEnsureProperStorageLocation()
            Using workspace As TestWorkspace = CreateWorkspace()
                AddHandler workspace.GlobalOptions.OptionChanged, Sub(sender, e)

                                                                      If SkippedOptionsForProperStorageTest.Contains(e.Option.Name) Then
                                                                          Return
                                                                      End If

                                                                      Assert.True(e.Option.StorageLocations.Any(
                                                                                  Function(s) TypeOf s Is RoamingProfileStorageLocation OrElse TypeOf s Is LocalUserProfileStorageLocation OrElse TypeOf s Is FeatureFlagStorageLocation),
                                                                                  $"Option '{e.Option.Name}' doesn't have proper storage location for persistence.")
                                                                  End Sub

                Dim automationObject As AbstractAutomationObject = CreateAutomationObject(workspace)
                Dim automationObjectType = automationObject.GetType()

                For Each [property] In automationObjectType.GetProperties()
                    Assert.True([property].CanRead, $"'{[property].Name}' must have a getter.")
                    Assert.True([property].CanWrite, $"'{[property].Name}' must have a setter.")

                    If SkippedOptionsForOptionChangedTest.Contains([property].Name) Then
                        Continue For
                    End If

                    Dim defaultValue = [property].GetValue(automationObject, Nothing)
                    Dim newValue As Object = GetNewValue(defaultValue, [property].Name)

                    Assert.True(newValue IsNot defaultValue, $"'{[property].Name}' must have a different newValue.")
                    Assert.Raises(Of OptionChangedEventArgs)(attach:=Sub(h) AddHandler workspace.GlobalOptions.OptionChanged, h,
                                                             detach:=Sub(h) RemoveHandler workspace.GlobalOptions.OptionChanged, h,
                                                             testCode:=Sub() [property].SetValue(automationObject, newValue))

                    Dim retrievedNewValue = [property].GetValue(automationObject, Nothing)
                    Assert.True(retrievedNewValue.Equals(newValue), $"'{[property].Name}' didn't retrieve the new value correctly.")
                Next
            End Using
        End Sub

        ' TODO: https://github.com/dotnet/roslyn/issues/62384
        Protected Shared Sub VerifySingleChangeWhenOptionChangeInUI(automationObject As AbstractAutomationObject, changeUIControl As Action, optionService As Object, optionStore As OptionStore, optionForAssertMessage As String)
            Dim automationValuesBeforeChange = GetAutomationDictionary(automationObject)

            changeUIControl()

            ' Following simulates the SaveSettingsToStorage call.
            ' Save the changes that were accumulated in the option store.
            Dim oldOptions As New SolutionOptionSet(DirectCast(optionService, ILegacyWorkspaceOptionService))
            Dim newOptions = DirectCast(optionStore.GetOptions(), SolutionOptionSet)

            ' Must log the option change before setting the new option values via s_optionService,
            ' otherwise oldOptions and newOptions would be identical and nothing will be logged.
            OptionLogger.Log(oldOptions, newOptions)
            DirectCast(optionService, ILegacyWorkspaceOptionService).SetOptions(newOptions, newOptions.GetChangedOptions())

            Dim automationValuesAfterChange = GetAutomationDictionary(automationObject)

            Assert.Equal(automationValuesBeforeChange.Count, automationValuesAfterChange.Count)
            AssertExactlyOneChange(automationValuesBeforeChange, automationValuesAfterChange, optionForAssertMessage)
        End Sub

        Private Shared Function GetAutomationDictionary(automationObject As AbstractAutomationObject) As ImmutableDictionary(Of String, Object)
            Dim automationObjectValues = ImmutableDictionary.CreateBuilder(Of String, Object)()
            For Each [property] In automationObject.GetType().GetProperties()
                automationObjectValues.Add([property].Name, [property].GetValue(automationObject, Nothing))
            Next

            Return automationObjectValues.ToImmutable()
        End Function

        Private Shared Sub AssertExactlyOneChange(dictionary1 As ImmutableDictionary(Of String, Object), dictionary2 As ImmutableDictionary(Of String, Object), optionForAssertMessage As String)
            Dim seenChange = False
            Dim changedKey As String = Nothing
            For Each key In dictionary1.Keys
                If Not dictionary1(key).Equals(dictionary2(key)) Then
                    If seenChange Then
                        If key <> "ShowSnippets" AndAlso changedKey <> "ShowSnippets" AndAlso
                            key <> "EnterKeyBehavior" AndAlso changedKey <> "EnterKeyBehavior" Then
                            Assert.False(True, $"Two values ('{key}' and '{changedKey}') have changed in automation object after changing '{optionForAssertMessage}'!")
                        End If
                    End If

                    changedKey = key
                    seenChange = True
                End If
            Next

            Assert.True(seenChange, $"No change was found in automation object after changing '{optionForAssertMessage}'.")
        End Sub

        <WpfFact>
        Public Sub TestOptionsInUIShouldBeInAutomationObject()
            Using workspace = CreateWorkspace(EditorTestCompositions.LanguageServerProtocol.AddExcludedPartTypes(GetType(ColorSchemeApplier)).AddParts(GetType(TestColorSchemeApplier)))
                Dim optionStore As New OptionStore(workspace.Options, Enumerable.Empty(Of IOption)())
                Dim optionService = workspace.Services.GetRequiredService(Of ILegacyWorkspaceOptionService)()
                Dim automationObject = CreateAutomationObject(workspace)

                Dim pageControls As IEnumerable(Of AbstractOptionPageControl) = CreatePageControls(optionStore, workspace)
                For Each pageControl In pageControls
                    Dim radioButtonGroups = New Dictionary(Of String, List(Of RadioButton))()
                    For Each bindingExpression In pageControl.GetTestAccessor().BindingExpressions
                        Dim target = bindingExpression.Target
                        Dim optionForAssertMessage = DirectCast(target, FrameworkElement).Name

                        If TypeOf target Is CheckBox Then
                            Dim checkBox = DirectCast(target, CheckBox)
                            VerifySingleChangeWhenOptionChangeInUI(automationObject, Sub() checkBox.IsChecked = Not checkBox.IsChecked, optionService, optionStore, optionForAssertMessage)
                        ElseIf TypeOf target Is ComboBox Then
                            Dim comboBox = DirectCast(target, ComboBox)
                            VerifySingleChangeWhenOptionChangeInUI(automationObject, Sub() comboBox.SelectedIndex = If(comboBox.SelectedIndex = 0, 1, 0), optionService, optionStore, optionForAssertMessage)
                        ElseIf TypeOf target Is RadioButton Then
                            Dim radioButton = DirectCast(target, RadioButton)
                            Dim list As List(Of RadioButton) = Nothing
                            If radioButtonGroups.TryGetValue(radioButton.GroupName, list) Then
                                list.Add(radioButton)
                            Else
                                radioButtonGroups.Add(radioButton.GroupName, New List(Of RadioButton) From {radioButton})
                            End If

                            Continue For
                        End If
                    Next

                    For Each radioButtonGroup In radioButtonGroups
                        Dim groupName = radioButtonGroup.Key
                        Dim radioButtons = radioButtonGroup.Value
                        ' There is no point in having a single radio button in a group.
                        Assert.True(radioButtons.Count > 1, $"Expected radio button group '{groupName}' to have more than one radio button. Found {radioButtons.Count}.")
                        Dim selectedRadioButton = radioButtons.SingleOrDefault(Function(r) r.IsChecked.HasValue AndAlso r.IsChecked.Value)
                        For Each radioButton In radioButtons
                            ' We test selecting every radio button in the group.
                            ' We skip the already selected one till we are sure we tested other radio buttons.
                            If radioButton Is selectedRadioButton Then
                                Continue For
                            End If

                            Assert.False(radioButton.IsChecked)
                            VerifySingleChangeWhenOptionChangeInUI(automationObject, Sub() radioButton.IsChecked = True, optionService, optionStore, optionForAssertMessage:=radioButton.Name)
                        Next

                        ' TODO: Consider asserting a non-null selectedRadioButton if https://github.com/dotnet/roslyn/issues/62363 is fixed.

                        If selectedRadioButton IsNot Nothing Then
                            ' Now that we tested other radio buttons in the group, the initially selected radio button is now not selected.
                            Assert.False(selectedRadioButton.IsChecked)
                            VerifySingleChangeWhenOptionChangeInUI(automationObject, Sub() selectedRadioButton.IsChecked = True, optionService, optionStore, optionForAssertMessage:=selectedRadioButton.Name)
                        End If
                    Next
                Next

                ' Above checks that all options are in AutomationObjects.
                ' TODO: check that all automation object members are in options.
            End Using
        End Sub

        Protected MustOverride Function GetNewValue(value As Object, propertyName As String) As Object
        Protected MustOverride Function CreateAutomationObject(workspace As TestWorkspace) As AbstractAutomationObject
        Protected MustOverride Function CreateWorkspace(Optional composition As TestComposition = Nothing) As TestWorkspace
        Protected MustOverride Function CreatePageControls(optionStore As OptionStore, workspace As TestWorkspace) As IEnumerable(Of AbstractOptionPageControl)

        <Export(GetType(IColorSchemeApplier)), [Shared], PartNotDiscoverable>
        Private Class TestColorSchemeApplier
            Implements IColorSchemeApplier

            <ImportingConstructor>
            <Obsolete(MefConstruction.ImportingConstructorMessage, True)>
            Public Sub New()
            End Sub

            Public Function IsSupportedThemeAsync(cancellationToken As CancellationToken) As Task(Of Boolean) Implements IColorSchemeApplier.IsSupportedThemeAsync
                Return Task.FromResult(True)
            End Function

            Public Function IsThemeCustomizedAsync(cancellationToken As CancellationToken) As Task(Of Boolean) Implements IColorSchemeApplier.IsThemeCustomizedAsync
                Return Task.FromResult(True)
            End Function
        End Class
    End Class
End Namespace
